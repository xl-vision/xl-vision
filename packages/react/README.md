# `@xl-vision/react`

> TODO: description

