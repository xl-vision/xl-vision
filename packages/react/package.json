{
  "name": "@xl-vision/react",
  "version": "0.0.0",
  "description": "xl-vision react components.",
  "author": "Rhys Xia <xrs4433@outlook.com>",
  "license": "MIT",
  "main": "legacy/index.js",
  "module": "modern/index.js",
  "types": "legacy/index.d.ts",
  "files": [
    "legacy",
    "modern"
  ],
  "keywords": [
    "xl-vision",
    "react",
    "styles"
  ],
  "homepage": "https://github.com/xl-vision/xl-vision-next/tree/master/packages/react#readme",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/xl-vision/xl-vision-next.git"
  },
  "scripts": {
    "build:legacy": "node ../../scripts/build legacy",
    "build:modern": "node ../../scripts/build modern",
    "build": "npm run build:legacy && npm run build:modern"
  },
  "bugs": {
    "url": "https://github.com/xl-vision/xl-vision-next/issues"
  },
  "dependencies": {
    "@babel/runtime": "^7.12.5",
    "prop-types": "^15.7.2"
  },
  "devDependencies": {
    "@types/react": "^16.8.0",
    "@types/react-dom": "^16.8.0",
    "@xl-vision/styled-engine": "^0.0.0",
    "react": "^16.8.0",
    "react-dom": "^16.8.0"
  },
  "peerDependencies": {
    "@types/react": "^16.8.0 || ^17.0.0",
    "@xl-vision/styled-engine": "^0.0.0",
    "react": "^16.8.0 || ^17.0.0",
    "react-dom": "^16.8.0 || ^17.0.0"
  },
  "peerDependenciesMeta": {
    "@types/react": {
      "optional": true
    }
  },
  "sideEffects": false,
  "publishConfig": {
    "access": "public"
  }
}
