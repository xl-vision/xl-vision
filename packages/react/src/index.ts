export { styled, createGlobalStyles } from './styles';
export { default as ThemeProvider } from './ThemeProvider';
export { default as Transition } from './Transition';
export { default as CSSTransition } from './CSSTransition';
export { default as CollapseTransition } from './CollapseTransition';
export { default as TransitionGroup } from './TransitionGroup';
