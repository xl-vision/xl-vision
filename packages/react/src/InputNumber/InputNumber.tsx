import { useConstantFn, useValueChange } from '@xl-vision/hooks';
import { CaretDownOutlined, CaretUpOutlined } from '@xl-vision/icons';
import { isProduction, omit } from '@xl-vision/utils';
import { forwardRef, useState, FocusEventHandler, useEffect } from 'react';
import { Input, InputProps } from '../Input';
import { styled } from '../styles';

export type InputNumberValueType = number | null;

export type InputNumberProps = Omit<
  InputProps,
  | 'type'
  | 'onChange'
  | 'value'
  | 'defaultValue'
  | 'suffix'
  | 'allowClear'
  | 'showCount'
  | 'maxLength'
> & {
  onChange?: (value: InputNumberValueType) => void;
  value?: InputNumberValueType;
  defaultValue?: InputNumberValueType;
  min?: number;
  max?: number;
  parser?: (value: string) => InputNumberValueType;
  formatter?: (value: InputNumberValueType) => string;
  step?: number;
  precision?: number;
};

const displayName = 'InputNumber';

const InputNumberRoot = styled(Input, {
  name: displayName,
  slot: 'Root',
})(({ theme: { clsPrefix, size: themeSize } }) => {
  return {
    [`&.${clsPrefix}-input--focused`]: {
      [`.${clsPrefix}-input-number__controls`]: {
        opacity: 1,
      },
    },
    [`.${clsPrefix}-input__suffix`]: {
      margin: `-${themeSize.padding.y}px -${themeSize.padding.x}px -${themeSize.padding.y}px 0`,
    },
  };
});

const InputNumberControls = styled('span', {
  name: displayName,
  slot: 'Controls',
})(({ theme: { size: themeSize, colors } }) => {
  return {
    opacity: 0,
    display: 'flex',
    height: '100%',
    borderLeft: `${themeSize.border}px solid ${colors.divider.primary}`,
    flexDirection: 'column',
  };
});

const InputNumberControlUp = styled('span', {
  name: displayName,
  slot: 'ControlUp',
})(({ theme: { transitions } }) => {
  return {
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'center',
    cursor: 'pointer',
    height: '40%',
    flex: 'auto',
    fontSize: '10px',
    padding: `0 3px`,
    transition: transitions.standard('all'),
    '&:hover': {
      height: '60%',
    },
  };
});

const InputNumberControlDown = styled(InputNumberControlUp, {
  name: displayName,
  slot: 'ControlDown',
})(({ theme: { size: themeSize, colors } }) => {
  return {
    borderTop: `${themeSize.border}px solid ${colors.divider.primary}`,
  };
});

const NUMBER_REGEX = /^-?\d+(\.\d+)?$/;

const InputNumber = forwardRef<HTMLSpanElement, InputNumberProps>((props, ref) => {
  const {
    onChange,
    value: valueProp,
    defaultValue = null,
    min,
    max,
    parser,
    formatter,
    onBlur,
    step = 1,
    precision,
    readOnly,
    disabled,
    ...others
  } = omit(
    props as InputProps,
    'allowClear',
    'showCount',
    'suffix',
    'maxLength',
  ) as InputNumberProps;

  const [value, setValue] = useValueChange(defaultValue, valueProp, onChange);

  const [internalValue, setInternalValue] = useState('');

  const handleFormatter = useConstantFn((v: InputNumberValueType) => {
    if (formatter) {
      return formatter(v);
    }
    if (v === null) {
      return '';
    }

    if (precision) {
      return v.toFixed(precision);
    }

    return String(v);
  });

  const defaultParser = useConstantFn((str: string) => {
    // eslint-disable-next-line react/destructuring-assignment
    const trimedValue = str.trim();
    if (!trimedValue) {
      return null;
    }
    if (NUMBER_REGEX.test(trimedValue)) {
      return Number(trimedValue);
    }
    return value;
  });

  const handleParser = useConstantFn((str: string) => {
    if (parser) {
      const parseredValue = parser(str);

      if (typeof parseredValue === 'string') {
        return defaultParser(parseredValue);
      }

      if (parseredValue !== null && typeof parseredValue !== 'number') {
        return value;
      }
      return parseredValue;
    }

    const v = defaultParser(str);

    if (v === null) {
      return v;
    }

    return v;
  });

  const updateValue = useConstantFn((newValue: InputNumberValueType | string) => {
    let v = typeof newValue === 'string' ? handleParser(newValue) : newValue;

    if (v !== null && precision !== undefined) {
      const magnification = 10 ** precision;

      v = Math.round(v * magnification) / magnification;
    }

    if (v !== null) {
      if (max !== undefined && v > max) {
        v = value;
      } else if (min !== undefined && v < min) {
        v = value;
      }
    }

    setValue(v);
    return v;
  });

  useEffect(() => {
    setInternalValue(handleFormatter(value));
  }, [value, handleFormatter]);

  const handleBlur: FocusEventHandler<HTMLInputElement> = useConstantFn((e) => {
    onBlur?.(e);
    const v = updateValue(internalValue);
    if (v === value) {
      // 有可能value没有变化，导致internalValue也不会更新
      setInternalValue(handleFormatter(v));
    }
  });

  const handleInternalValueChange = useConstantFn((str: string) => {
    setInternalValue(str);
    updateValue(str);
  });

  // TODO: remove BigInt
  const handleUp = useConstantFn(() => {
    const ret = BigInt(value || 0) + BigInt(step);

    updateValue(ret.toString());
  });

  // TODO: remove BigInt
  const handleDown = useConstantFn(() => {
    const ret = BigInt(value || 0) - BigInt(step);

    updateValue(ret.toString());
  });

  const suffixNode = !readOnly && !disabled && (
    <InputNumberControls>
      <InputNumberControlUp onClick={handleUp}>
        <CaretUpOutlined />
      </InputNumberControlUp>
      <InputNumberControlDown onClick={handleDown}>
        <CaretDownOutlined />
      </InputNumberControlDown>
    </InputNumberControls>
  );

  return (
    <InputNumberRoot
      disabled={disabled}
      readOnly={readOnly}
      ref={ref}
      suffix={suffixNode}
      value={internalValue}
      onBlur={handleBlur}
      onChange={handleInternalValueChange}
      {...others}
    />
  );
});

if (!isProduction) {
  InputNumber.displayName = displayName;
  InputNumber.propTypes = {};
}

export default InputNumber;
