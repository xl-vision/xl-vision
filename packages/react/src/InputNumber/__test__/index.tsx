import { fireEvent, render } from '@testing-library/react';
import InputNumber from '../InputNumber';

describe('InputNumber', () => {
  it('Test onChange', () => {
    const fn = jest.fn();

    const wrapper = render(<InputNumber defaultValue={5} onChange={fn} />);

    expect(fn.mock.calls.length).toBe(0);

    const input = wrapper.container.querySelector('input')!;

    fireEvent.change(input, {
      target: {
        value: '6',
      },
    });

    expect(fn.mock.calls.length).toBe(1);
    expect(fn.mock.calls[0][0]).toBe(6);
    fn.mockClear();

    fireEvent.change(input, {
      target: {
        value: '6',
      },
    });

    expect(fn.mock.calls.length).toBe(0);
  });
});
