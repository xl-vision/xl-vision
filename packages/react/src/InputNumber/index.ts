import InputNumber from './InputNumber';

export { default as InputNumber } from './InputNumber';
export * from './InputNumber';

export default InputNumber;
