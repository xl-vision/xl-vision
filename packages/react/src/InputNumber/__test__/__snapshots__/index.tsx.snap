// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputNumber Test controls 1`] = `
.emotion-0 {
  font-size: 1.000rem;
  line-height: 1.5;
  font-weight: 400;
  letter-spacing: 0.031em;
  -webkit-text-decoration: inherit;
  text-decoration: inherit;
  text-transform: inherit;
  font-family: 'Roboto','Helvetica Neue','segoe ui','Arial','noto sans',sans-serif;
  width: 100%;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
}

.emotion-0 .xl-input__suffix {
  margin: -4px -10px -4px 0;
  -webkit-transform: scaleX(0);
  -moz-transform: scaleX(0);
  -ms-transform: scaleX(0);
  transform: scaleX(0);
  transform-origin: 100% 50%;
  -webkit-transition: -webkit-transform 300ms 0ms cubic-bezier(0, 0, 0.2, 1);
  transition: transform 300ms 0ms cubic-bezier(0, 0, 0.2, 1);
  overflow: hidden;
}

.emotion-0:hover .xl-input__suffix,
.emotion-0.xl-input--focused .xl-input__suffix {
  -webkit-transform: scaleX(1);
  -moz-transform: scaleX(1);
  -ms-transform: scaleX(1);
  transform: scaleX(1);
}

.emotion-1 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  border-radius: 4px;
  border: 1px solid #d9d9d9;
  width: 100%;
  padding: 4px 10px;
  color: #000000e0;
  background-color: #fff;
  -webkit-transition: border-color 300ms 0ms cubic-bezier(0.4, 0, 0.2, 1),box-shadow 300ms 0ms cubic-bezier(0.4, 0, 0.2, 1);
  transition: border-color 300ms 0ms cubic-bezier(0.4, 0, 0.2, 1),box-shadow 300ms 0ms cubic-bezier(0.4, 0, 0.2, 1);
  z-index: 1;
  font-size: inherit;
}

.emotion-1:not(:first-child) {
  border-top-left-radius: 0;
  border-bottom-left-radius: 0;
}

.emotion-1:not(:last-child) {
  border-top-right-radius: 0;
  border-bottom-right-radius: 0;
}

.emotion-1:hover {
  border-color: #60a5fa;
}

.emotion-2 {
  font-size: inherit;
  line-height: 1.5;
  font-weight: 400;
  letter-spacing: 0.031em;
  -webkit-text-decoration: inherit;
  text-decoration: inherit;
  text-transform: inherit;
  font-family: 'Roboto','Helvetica Neue','segoe ui','Arial','noto sans',sans-serif;
  cursor: inherit;
  touch-action: manipulation;
  font-variant: tabular-nums;
  display: inline-block;
  min-width: 0;
  width: 100%;
  border: 0;
  outline: 0;
  padding: 0;
  -webkit-appearance: none;
  color: #000000e0;
  background-color: #fff;
  -webkit-transition: border-color 300ms 0ms cubic-bezier(0.4, 0, 0.2, 1),box-shadow 300ms 0ms cubic-bezier(0.4, 0, 0.2, 1);
  transition: border-color 300ms 0ms cubic-bezier(0.4, 0, 0.2, 1),box-shadow 300ms 0ms cubic-bezier(0.4, 0, 0.2, 1);
}

.emotion-2::-moz-placeholder {
  opacity: 1;
}

.emotion-2::-webkit-input-placeholder {
  color: #00000061;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.emotion-2::-moz-placeholder {
  color: #00000061;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.emotion-2:-ms-input-placeholder {
  color: #00000061;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.emotion-2::placeholder {
  color: #00000061;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.emotion-2:placeholder-shown {
  text-overflow: ellipsis;
}

.emotion-3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: none;
  -ms-flex: none;
  flex: none;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  margin-right: 4px;
  margin-right: 0;
  margin-left: 4px;
}

.emotion-3 .xl-input__suffix--has-suffix {
  margin-right: 4px;
}

.emotion-3 .xl-input__suffix-count,
.emotion-3 .xl-input__suffix-clear {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  line-height: 1;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  color: #00000061;
}

.emotion-3 .xl-input__suffix-clear {
  cursor: pointer;
  -webkit-transition: color 300ms 0ms cubic-bezier(0.4, 0, 0.2, 1);
  transition: color 300ms 0ms cubic-bezier(0.4, 0, 0.2, 1);
}

.emotion-3 .xl-input__suffix-clear:hover {
  color: #000000a6;
}

.emotion-4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 100%;
  border-left: 1px solid #d9d9d9;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.emotion-5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  cursor: pointer;
  height: 40%;
  -webkit-flex: auto;
  -ms-flex: auto;
  flex: auto;
  font-size: 10px;
  padding: 0 3px;
  -webkit-transition: all 300ms 0ms cubic-bezier(0.4, 0, 0.2, 1);
  transition: all 300ms 0ms cubic-bezier(0.4, 0, 0.2, 1);
}

.emotion-5:hover {
  height: 60%;
}

.emotion-5.xl-input-number__control--disabled {
  cursor: not-allowed;
  opacity: 0.3;
}

.emotion-5.xl-input-number__control--disabled:hover {
  height: 40%;
}

.emotion-6 {
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  width: 1em;
  height: 1em;
  display: inline-block;
  fill: currentcolor;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
}

.emotion-7 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  cursor: pointer;
  height: 40%;
  -webkit-flex: auto;
  -ms-flex: auto;
  flex: auto;
  font-size: 10px;
  padding: 0 3px;
  -webkit-transition: all 300ms 0ms cubic-bezier(0.4, 0, 0.2, 1);
  transition: all 300ms 0ms cubic-bezier(0.4, 0, 0.2, 1);
  border-top: 1px solid #d9d9d9;
}

.emotion-7:hover {
  height: 60%;
}

.emotion-7.xl-input-number__control--disabled {
  cursor: not-allowed;
  opacity: 0.3;
}

.emotion-7.xl-input-number__control--disabled:hover {
  height: 40%;
}

<div>
  <span
    class="xl-input xl-input--size-middle xl-input-number emotion-0"
  >
    <span
      class="xl-input__wrapper emotion-1"
    >
      <input
        autocomplete="off"
        class="xl-input__inner emotion-2"
        role="spinbutton"
        step="1"
        type="text"
        value=""
      />
      <span
        class="xl-input__prefix xl-input__suffix emotion-3"
      >
        <span
          class="xl-input-number__controls emotion-4"
        >
          <span
            class="xl-input-number__control-up emotion-5"
          >
            <svg
              class="xl-base-icon emotion-6"
              viewBox="0 0 1024 1024"
            >
              <path
                d="M858.9 689 530.5 308.2c-9.4-10.9-27.5-10.9-37 0L165.1 689c-12.2 14.2-1.2 35 18.5 35h656.8c19.7 0 30.7-20.8 18.5-35"
              />
            </svg>
          </span>
          <span
            class="xl-input-number__control-up xl-input-number__control-down emotion-7"
          >
            <svg
              class="xl-base-icon emotion-6"
              viewBox="0 0 1024 1024"
            >
              <path
                d="M840.4 300H183.6c-19.7 0-30.7 20.8-18.5 35l328.4 380.8c9.4 10.9 27.5 10.9 37 0L858.9 335c12.2-14.2 1.2-35-18.5-35"
              />
            </svg>
          </span>
        </span>
      </span>
    </span>
  </span>
</div>
`;

exports[`InputNumber Test controls 2`] = `
.emotion-0 {
  font-size: 1.000rem;
  line-height: 1.5;
  font-weight: 400;
  letter-spacing: 0.031em;
  -webkit-text-decoration: inherit;
  text-decoration: inherit;
  text-transform: inherit;
  font-family: 'Roboto','Helvetica Neue','segoe ui','Arial','noto sans',sans-serif;
  width: 100%;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
}

.emotion-0 .xl-input__suffix {
  margin: -4px -10px -4px 0;
  -webkit-transform: scaleX(0);
  -moz-transform: scaleX(0);
  -ms-transform: scaleX(0);
  transform: scaleX(0);
  transform-origin: 100% 50%;
  -webkit-transition: -webkit-transform 300ms 0ms cubic-bezier(0, 0, 0.2, 1);
  transition: transform 300ms 0ms cubic-bezier(0, 0, 0.2, 1);
  overflow: hidden;
}

.emotion-0:hover .xl-input__suffix,
.emotion-0.xl-input--focused .xl-input__suffix {
  -webkit-transform: scaleX(1);
  -moz-transform: scaleX(1);
  -ms-transform: scaleX(1);
  transform: scaleX(1);
}

.emotion-1 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  border-radius: 4px;
  border: 1px solid #d9d9d9;
  width: 100%;
  padding: 4px 10px;
  color: #000000e0;
  background-color: #fff;
  -webkit-transition: border-color 300ms 0ms cubic-bezier(0.4, 0, 0.2, 1),box-shadow 300ms 0ms cubic-bezier(0.4, 0, 0.2, 1);
  transition: border-color 300ms 0ms cubic-bezier(0.4, 0, 0.2, 1),box-shadow 300ms 0ms cubic-bezier(0.4, 0, 0.2, 1);
  z-index: 1;
  font-size: inherit;
}

.emotion-1:not(:first-child) {
  border-top-left-radius: 0;
  border-bottom-left-radius: 0;
}

.emotion-1:not(:last-child) {
  border-top-right-radius: 0;
  border-bottom-right-radius: 0;
}

.emotion-1:hover {
  border-color: #60a5fa;
}

.emotion-2 {
  font-size: inherit;
  line-height: 1.5;
  font-weight: 400;
  letter-spacing: 0.031em;
  -webkit-text-decoration: inherit;
  text-decoration: inherit;
  text-transform: inherit;
  font-family: 'Roboto','Helvetica Neue','segoe ui','Arial','noto sans',sans-serif;
  cursor: inherit;
  touch-action: manipulation;
  font-variant: tabular-nums;
  display: inline-block;
  min-width: 0;
  width: 100%;
  border: 0;
  outline: 0;
  padding: 0;
  -webkit-appearance: none;
  color: #000000e0;
  background-color: #fff;
  -webkit-transition: border-color 300ms 0ms cubic-bezier(0.4, 0, 0.2, 1),box-shadow 300ms 0ms cubic-bezier(0.4, 0, 0.2, 1);
  transition: border-color 300ms 0ms cubic-bezier(0.4, 0, 0.2, 1),box-shadow 300ms 0ms cubic-bezier(0.4, 0, 0.2, 1);
}

.emotion-2::-moz-placeholder {
  opacity: 1;
}

.emotion-2::-webkit-input-placeholder {
  color: #00000061;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.emotion-2::-moz-placeholder {
  color: #00000061;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.emotion-2:-ms-input-placeholder {
  color: #00000061;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.emotion-2::placeholder {
  color: #00000061;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.emotion-2:placeholder-shown {
  text-overflow: ellipsis;
}

.emotion-3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: none;
  -ms-flex: none;
  flex: none;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  margin-right: 4px;
  margin-right: 0;
  margin-left: 4px;
}

.emotion-3 .xl-input__suffix--has-suffix {
  margin-right: 4px;
}

.emotion-3 .xl-input__suffix-count,
.emotion-3 .xl-input__suffix-clear {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  line-height: 1;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  color: #00000061;
}

.emotion-3 .xl-input__suffix-clear {
  cursor: pointer;
  -webkit-transition: color 300ms 0ms cubic-bezier(0.4, 0, 0.2, 1);
  transition: color 300ms 0ms cubic-bezier(0.4, 0, 0.2, 1);
}

.emotion-3 .xl-input__suffix-clear:hover {
  color: #000000a6;
}

<div>
  <span
    class="xl-input xl-input--size-middle xl-input-number emotion-0"
  >
    <span
      class="xl-input__wrapper emotion-1"
    >
      <input
        autocomplete="off"
        class="xl-input__inner emotion-2"
        role="spinbutton"
        step="1"
        type="text"
        value=""
      />
      <span
        class="xl-input__prefix xl-input__suffix emotion-3"
      />
    </span>
  </span>
</div>
`;

exports[`InputNumber Test controls 3`] = `
.emotion-0 {
  font-size: 1.000rem;
  line-height: 1.5;
  font-weight: 400;
  letter-spacing: 0.031em;
  -webkit-text-decoration: inherit;
  text-decoration: inherit;
  text-transform: inherit;
  font-family: 'Roboto','Helvetica Neue','segoe ui','Arial','noto sans',sans-serif;
  width: 100%;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
}

.emotion-0 .xl-input__suffix {
  margin: -4px -10px -4px 0;
  -webkit-transform: scaleX(0);
  -moz-transform: scaleX(0);
  -ms-transform: scaleX(0);
  transform: scaleX(0);
  transform-origin: 100% 50%;
  -webkit-transition: -webkit-transform 300ms 0ms cubic-bezier(0, 0, 0.2, 1);
  transition: transform 300ms 0ms cubic-bezier(0, 0, 0.2, 1);
  overflow: hidden;
}

.emotion-0:hover .xl-input__suffix,
.emotion-0.xl-input--focused .xl-input__suffix {
  -webkit-transform: scaleX(1);
  -moz-transform: scaleX(1);
  -ms-transform: scaleX(1);
  transform: scaleX(1);
}

.emotion-1 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  border-radius: 4px;
  border: 1px solid #d9d9d9;
  width: 100%;
  padding: 4px 10px;
  color: #000000e0;
  background-color: #fff;
  -webkit-transition: border-color 300ms 0ms cubic-bezier(0.4, 0, 0.2, 1),box-shadow 300ms 0ms cubic-bezier(0.4, 0, 0.2, 1);
  transition: border-color 300ms 0ms cubic-bezier(0.4, 0, 0.2, 1),box-shadow 300ms 0ms cubic-bezier(0.4, 0, 0.2, 1);
  z-index: 1;
  font-size: inherit;
}

.emotion-1:not(:first-child) {
  border-top-left-radius: 0;
  border-bottom-left-radius: 0;
}

.emotion-1:not(:last-child) {
  border-top-right-radius: 0;
  border-bottom-right-radius: 0;
}

.emotion-1:hover {
  border-color: #60a5fa;
}

.emotion-2 {
  font-size: inherit;
  line-height: 1.5;
  font-weight: 400;
  letter-spacing: 0.031em;
  -webkit-text-decoration: inherit;
  text-decoration: inherit;
  text-transform: inherit;
  font-family: 'Roboto','Helvetica Neue','segoe ui','Arial','noto sans',sans-serif;
  cursor: inherit;
  touch-action: manipulation;
  font-variant: tabular-nums;
  display: inline-block;
  min-width: 0;
  width: 100%;
  border: 0;
  outline: 0;
  padding: 0;
  -webkit-appearance: none;
  color: #000000e0;
  background-color: #fff;
  -webkit-transition: border-color 300ms 0ms cubic-bezier(0.4, 0, 0.2, 1),box-shadow 300ms 0ms cubic-bezier(0.4, 0, 0.2, 1);
  transition: border-color 300ms 0ms cubic-bezier(0.4, 0, 0.2, 1),box-shadow 300ms 0ms cubic-bezier(0.4, 0, 0.2, 1);
}

.emotion-2::-moz-placeholder {
  opacity: 1;
}

.emotion-2::-webkit-input-placeholder {
  color: #00000061;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.emotion-2::-moz-placeholder {
  color: #00000061;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.emotion-2:-ms-input-placeholder {
  color: #00000061;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.emotion-2::placeholder {
  color: #00000061;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.emotion-2:placeholder-shown {
  text-overflow: ellipsis;
}

.emotion-3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: none;
  -ms-flex: none;
  flex: none;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  margin-right: 4px;
  margin-right: 0;
  margin-left: 4px;
}

.emotion-3 .xl-input__suffix--has-suffix {
  margin-right: 4px;
}

.emotion-3 .xl-input__suffix-count,
.emotion-3 .xl-input__suffix-clear {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  line-height: 1;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  color: #00000061;
}

.emotion-3 .xl-input__suffix-clear {
  cursor: pointer;
  -webkit-transition: color 300ms 0ms cubic-bezier(0.4, 0, 0.2, 1);
  transition: color 300ms 0ms cubic-bezier(0.4, 0, 0.2, 1);
}

.emotion-3 .xl-input__suffix-clear:hover {
  color: #000000a6;
}

.emotion-4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 100%;
  border-left: 1px solid #d9d9d9;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.emotion-5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  cursor: pointer;
  height: 40%;
  -webkit-flex: auto;
  -ms-flex: auto;
  flex: auto;
  font-size: 10px;
  padding: 0 3px;
  -webkit-transition: all 300ms 0ms cubic-bezier(0.4, 0, 0.2, 1);
  transition: all 300ms 0ms cubic-bezier(0.4, 0, 0.2, 1);
}

.emotion-5:hover {
  height: 60%;
}

.emotion-5.xl-input-number__control--disabled {
  cursor: not-allowed;
  opacity: 0.3;
}

.emotion-5.xl-input-number__control--disabled:hover {
  height: 40%;
}

.emotion-6 {
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  width: 1em;
  height: 1em;
  display: inline-block;
  fill: currentcolor;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
}

.emotion-7 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  cursor: pointer;
  height: 40%;
  -webkit-flex: auto;
  -ms-flex: auto;
  flex: auto;
  font-size: 10px;
  padding: 0 3px;
  -webkit-transition: all 300ms 0ms cubic-bezier(0.4, 0, 0.2, 1);
  transition: all 300ms 0ms cubic-bezier(0.4, 0, 0.2, 1);
  border-top: 1px solid #d9d9d9;
}

.emotion-7:hover {
  height: 60%;
}

.emotion-7.xl-input-number__control--disabled {
  cursor: not-allowed;
  opacity: 0.3;
}

.emotion-7.xl-input-number__control--disabled:hover {
  height: 40%;
}

<div>
  <span
    class="xl-input xl-input--size-middle xl-input-number emotion-0"
  >
    <span
      class="xl-input__wrapper emotion-1"
    >
      <input
        autocomplete="off"
        class="xl-input__inner emotion-2"
        role="spinbutton"
        step="1"
        type="text"
        value=""
      />
      <span
        class="xl-input__prefix xl-input__suffix emotion-3"
      >
        <span
          class="xl-input-number__controls emotion-4"
        >
          <span
            class="xl-input-number__control-up emotion-5"
          >
            <svg
              class="xl-base-icon emotion-6"
              viewBox="0 0 1024 1024"
            >
              <path
                d="M858.9 689 530.5 308.2c-9.4-10.9-27.5-10.9-37 0L165.1 689c-12.2 14.2-1.2 35 18.5 35h656.8c19.7 0 30.7-20.8 18.5-35"
              />
            </svg>
          </span>
          <span
            class="xl-input-number__control-up xl-input-number__control-down emotion-7"
          >
            <svg
              class="xl-base-icon emotion-6"
              viewBox="0 0 1024 1024"
            >
              <path
                d="M840.4 300H183.6c-19.7 0-30.7 20.8-18.5 35l328.4 380.8c9.4 10.9 27.5 10.9 37 0L858.9 335c12.2-14.2 1.2-35-18.5-35"
              />
            </svg>
          </span>
        </span>
      </span>
    </span>
  </span>
</div>
`;
